src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs

```diff
@@ -0,0 +1,431 @@
+                {
+                    // 0x747A646174640000 = {0x74, 0x7A, 0x64, 0x61, 0x74, 0x64, 0x00, 0x00} = "tzdata[NUL][NUL]"
+                    var b = new StringBuilder(signature.Length);
+                    for (int i = 0; i < signature.Length; ++i)
+                    {
+                        b.Append(' ').Append(HexConverter.ToCharLower(signature[i]));
+                    }
+
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_BadTZHeader, TimeZoneFileName, b.ToString()));
+                }
+
+                AndroidTzDataHeader header;
+
+                header.indexOffset = TZif_ToInt32(buffer.Slice(12, 4));
+                header.dataOffset = TZif_ToInt32(buffer.Slice(16, 4));
+                header.finalOffset = TZif_ToInt32(buffer.Slice(20, 4));
+
+                return header;
+            }
+
+            private unsafe void ReadIndex(Stream fs, int indexOffset, int dataOffset)
+            {
+                int indexSize = dataOffset - indexOffset;
+                int entrySize = sizeof(AndroidTzDataEntry);
+                int entryCount = indexSize / entrySize;
+
+                _byteOffsets = new int[entryCount];
+                _ids = new string[entryCount];
+                _lengths = new int[entryCount];
+
+                for (int i = 0; i < entryCount; ++i)
+                {
+                    AndroidTzDataEntry entry = LoadEntryAt(fs, indexOffset + (entrySize*i));
+                    var p = (sbyte*)entry.id;
+
+                    _byteOffsets![i] = entry.byteOffset + dataOffset;
+                    _ids![i] = new string(p);
+                    _lengths![i] = entry.length;
+
+                    if (_lengths![i] < sizeof(AndroidTzDataHeader))
+                    {
+                        throw new InvalidOperationException(SR.InvalidOperation_BadIndexLength);
+                    }
+                }
+            }
+
+            private unsafe AndroidTzDataEntry LoadEntryAt(Stream fs, long position)
+            {
+                int size = sizeof(AndroidTzDataEntry);
+                Span<byte> entryBuffer = stackalloc byte[size];
+
+                fs.Position = position;
+
+                int bytesRead = 0;
+                int bytesLeft = size;
+
+                while (bytesLeft > 0)
+                {
+                    int b = fs.Read(entryBuffer);
+                    if (b == 0)
+                    {
+                        break;
+                    }
+
+                    bytesRead += b;
+                    bytesLeft -= b;
+                }
+
+                if (bytesLeft != 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_ReadTZError, _tzFilePath, position, size, bytesRead, size));
+                }
+
+                AndroidTzDataEntry entry;
+                for (int i = 0; i < 40; i++)
+                {
+                    entry.id[i] = entryBuffer[i];
```

Comment:
Can we just CopyTo here?

```C#
entryBuffer.Slice(0, 40).CopyTo(entry.id);
```

Although, copying the bytes to the `AndroidTzDataEntry`, just to copy them out into a string doesn't seem that great IMO. Why don't we change `AndroidTzDataEntry` to have a string, and then just create the string here, directly from the buffer? That's one less copy that needs to occur.

`entry.unused` doesn't seem to be used, so there is no need in filling it out.

So I think `AndroidTzDataEntry` could just look like:

```C#
            private struct AndroidTzDataEntry
            {
                public string id;
                public int byteOffset;
                public int length;
            }
```

And then we copy that information out into `_byteOffsets`, `_ids` and `_lengths`

Reply:
Obviously, if you do this, you need to change the buffer size to not use `sizeof(AndroidTzDataEntry)`.

Reply:
> And then we copy that information out into _byteOffsets, _ids and _lengths

We can have just one `AndroidTzDataEntry[]` array instead of 3 related arrays if we do this.

Reply:
Could we still use Array.BinarySearch then to find the `id`?

https://github.com/dotnet/runtime/pull/54845/files#diff-ca73c8aae858710155c705be97320256749a054ca52bfc16628fc2e8313b7433R408

```C#
            public byte[] GetTimeZoneData(string id)
            {
                int i = Array.BinarySearch(_ids!, id, StringComparer.Ordinal);
```

Reply:
Changing AndroidTzDataEntry `id` to be of type string instead of `fixed byte id[40]` is giving me some trouble converting the `Span<byte>` from `entryBuffer.Slice(0, 40)` into a string. I cannot convert the `Span<byte>` into `sbyte*` or `entryBuffer.Slice(0, 40).ToArray()` to `sbyte*`. 

The method still has the unsafe keyword but I don't fully understand the best way to get from `Span<byte>` to a string. 

I also agree that if we converted the 3 arrays into one array, then 
```
public string[] GetTimeZoneIds()
{
    return _ids!;
}
```
Would need to construct a `string[]` and manually iterate over the AndroidTzDataEntry[].
Given that, we could use Array.BinarySearch over `GetTimeZoneIds()`, but it seems more direct to keep the `_ids` array instead.

Reply:
> I don't fully understand the best way to get from Span<byte> to a string.

In this case, I would just call:

```
Encoding.UTF8.GetString(entryBuffer.Slice(0, 40))
```

Reply:
Ah nevermind, I'll use `Encoding.Utf8.GetString(entryBuffer.Slice(0, 40));` instead.

Reply:
One thing I noticed with the call to `GetString` is that given a span with the ascii string "UTC", you'll get:

`8584670000000000000000000000000000000000000`

And spot checking the `GetString` code, you'll get a string that has the value "UTC", but the length will be 40 as opposed to 3.  It appears when creating a string from encoding, it'll include nulls as part of the count because it's valid ascii.

Reply:
The tests are passing from setting the id to `Encoding.UTF8.GetString(entryBuffer.Slice(0, 40)).Split('\0')[0];`.